src/triage/component/postmodeling/contrast/model_evaluator.py

```diff
@@ -0,0 +1,1155 @@
+        Y = [(np.where(condition, 1, -1), label) for label, condition in \
+             dict_error_class.items()]
+
+        # Define feature space to model: get the list of feature names
+        storage = ProjectStorage(path)
+        matrix_storage = MatrixStorageEngine(storage).get_store(self.pred_matrix_uuid)
+        feature_columns = matrix_storage.columns()
+
+        # Build error feature matrix
+        matrices = [test_matrix_thresh,
+                    test_matrix_thresh,
+                    test_matrix_thresh_1,
+                    test_matrix_thresh_0]
+        X = [matrix[feature_columns] for matrix in matrices]
+
+        return zip(Y, X)
+
+    def _error_modeler(self,
+                      depth=None,
+                      view_plots=False,
+                      **kwargs):
+       '''
+       Model labeled errors (residuals) by the error_labeler (FPR, FNR, and
+       general residual) using a RandomForestClassifier. This function will
+       yield a plot tree for each of the label numpy arrays return by the
+       error_labeler (Y).
+       Arguments:
+           - depth: max number of tree partitions. This is passed directly to
+             the classifier.
+           - view_plot: the plot is saved to disk by default, but the
+             graphviz.Source also allow to load the object and see it in the
+             default OS image renderer
+           - **kwargs: more arguments passed to the labeler: param indicating
+             the threshold value, param_type indicating the type of threshold,
+             and the path to the ProjectStorage.
+       '''
+
+       # Get matrices from the labeler
+       zip_data = self._error_labeler(param_type = kwargs['param_type'],
+                                  param = kwargs['param'],
+                                  path=kwargs['path'])
+
+       # Model tree and output tree plot
+       for error_label, matrix in zip_data:
+
+           dot_path = 'error_analysis_' + \
+                      str(error_label[1]) + '_' + \
+                      str(self.model_id) + '_' + \
+                      str(kwargs['param_type']) + '@'+ \
+                      str(kwargs['param']) +  '.gv'
+
+           clf = tree.DecisionTreeClassifier(max_depth=depth)
+           clf_fit = clf.fit(matrix, error_label[0])
+           tree_viz = tree.export_graphviz(clf_fit,
+                                           out_file=None,
+                                           feature_names=matrix.columns.values,
+                                           filled=True,
+                                           rounded=True,
+                                           special_characters=True)
+           graph = graphviz.Source(tree_viz)
+           graph.render(filename=dot_path,
+                        directory='error_analysis',
+                        view=view_plots)
+
+           print(dot_path)
+
+    def error_analysis(self, threshold_iterator, **kwargs):
```

Comment:
a note on naming -- i wanted to call the variable `threshold_iterator`, in my comment below, because it *is* an iterator, but I had to settle for `threshold_iterator_` -- because the object you're passing to this method is not *really* (just) an iterator.

In fact, it doesn't define the iterator interface at all, (*e.g.*: `__next__`).

However, your `ThresholdIterator` class does define the interface of an **iterable** -- something that defines its own iterator, via `__iter__`.

Generally, methods like these would indicate that they support **iterables**, (not merely **iterators**) -- for example, the parameter `threshold_iterable`.